src/index.js

```diff
@@ -0,0 +1,224 @@
+import { exec } from 'child_process';
+import { sync as findUp } from 'find-up';
+import * as fs from 'fs-extra';
+import globCb from 'glob';
+import * as path from 'path';
+import semver from 'semver';
+import * as zkSnark from 'snarkjs';
+import { promisify } from 'util';
+
+const glob = promisify(globCb);
+
+zkSnark.bigInt.prototype.toJSON = function() {
+  return this.toString();
+};
+
+const Extension = {
+  Json: '.json',
+  Circom: '.circom',
+  VkProof: '.vk_proof',
+  VkVerifier: '.vk_verifier',
+  Solidity: '.sol'
+};
+
+class Snarks {
+  constructor(embark) {
+    this.embark = embark;
+    this.circuitsConfig = embark.pluginConfig;
+    this.compiledCircuitsPath = embark.config.dappPath('.embark', 'snarks');
+    // v5 detection won't work as desired with "dev embark" (built in monorepo)
+    // until after first prerelease of embark v5; can assign `true` to test
+    // this.isEmbark5 = true;
+    this.isEmbark5 = semver(this.embark.version).major >= 5;
+    this.registerEvents();
+  }
+
+  static circomBinary = findUp('node_modules/.bin/circom', { cwd: __dirname });
+  static snarkjsBinary = findUp('node_modules/.bin/snarkjs', {
+    cwd: __dirname
+  });
+
+  registerEvents() {
+    if (this.isEmbark5) {
+      return this.embark.registerActionForEvent(
+        'compiler:contracts:compile:before',
+        (contractFiles, callback) => {
+          this.compileAndGenerateContracts(contractFiles, callback);
+        }
+      );
+    }
+    this.embark.registerActionForEvent('build:beforeAll', callback =>
+      this.compileAndGenerateContracts([], callback)
+    );
+  }
+
+  verifierFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkVerifier}`
+    );
+  }
+
+  verifierContractPath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.Solidity}`
+    );
+  }
+
+  proofFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkProof}`
+    );
+  }
+
+  async compileCircuits() {
+    await fs.ensureDir(this.compiledCircuitsPath);
+
+    const patterns = this.circuitsConfig.circuits;
+    if (!patterns) {
+      return;
+    }
+
+    this.embark.logger.info('Compiling circuits...');
+
+    await Promise.all(
+      patterns.map(async pattern => {
+        const filepaths = await glob(pattern);
+        await Promise.all(
+          filepaths
+            .filter(filename => path.extname(filename) === Extension.Circom)
+            .map(filepath => this.compileCircuit(filepath))
+        );
+      })
+    );
```

Comment:
Nice composition! 👍

Reply:
Thanks!  I can't take too much credit. 95% of the code in this file was written by Anthony. I made several things async that didn't need to be sync, and had to figure out the multi-version compatibility story w.r.t. embark.